### ISSUE TITLE
Cannot create a foreign key to a reference table from a partitioned table

### ISSUE THREAD

**Issue by mtuncer** [citusdata/citus#2592]
This has started to occur today due to a recent change in PG11. Travis was used to be fine until Jan 18th. Some change afterwards broke the master branch.

Smallest repro I could produce at the moment

```sql
DROP TABLE IF EXISTS partitioning_test, partitioning_test_2009, partitioning_test_2012, partitioning_test_reference CASCADE;


CREATE TABLE partitioning_test(id int, time date) PARTITION BY RANGE (time);
SELECT create_distributed_table('partitioning_test', 'id');

CREATE TABLE partitioning_test_2009 PARTITION OF partitioning_test FOR VALUES FROM ('2009-01-01') TO ('2010-01-01');

-- create a reference table for foreign key test
CREATE TABLE partitioning_test_reference(id int PRIMARY KEY, subid int);
SELECT create_reference_table('partitioning_test_reference');

ALTER TABLE partitioning_test ADD CONSTRAINT partitioning_reference_fkey FOREIGN KEY (id) REFERENCES partitioning_test_reference(id) ON DELETE CASCADE;
```

`ALTER TABLE` command fails with
```
ERROR:  cannot perform DDL on placement 3208, which has been read over multiple connections
```

Creating the foreign key from a partition to a reference table works fine. Problem is with the parent table. This works fine on PG10 but fails on PG11.

This is causing all PRs on travis to fail. We can't merge anything else unless we fix this.

We should also backport the fix to earlier versions that support PG11.

**Comment by mtuncer** [comment]
Looks like  this is a side effect of Alvaro Herrera's recent commit https://github.com/postgres/postgres/commit/0325d7a5957ba39a0dce90835ab54a08ab8bf762 

I haven't confirmed this yet

**Comment by mtuncer** [comment]
Findings so far
- issue occurs only if partitioned table has a partition that does not have the same foreign key created.
- PG tries to validate the foreign key by running a select query on the partition table and the reference to see if there is any violation.
- This query gets to Citus via SPI call when we let the standard_executor run during in process https://github.com/citusdata/citus/blob/master/src/backend/distributed/commands/utility_hook.c#L422
- We end up executing the select query, connection is marked as run a select query
- When we get the control back to actually perform the DDL operation. We detect that a select is already performed in this connection and error out.

We did not have this problem because, postgres did not propagate foreign key creation commands to partitions in the past.

Still thinking about how we should fix this.

**Comment by mtuncer** [comment]
Now we have an interim fix however this is not the final fix.

There is still some changes happening in the related code in PG side. We should keep this issue open and get a final fix.

Current fix is to set sequential execution when alter table with foreign keys to reference table. 

However, this does not fix attach table since foreign key validation occurs before the table shards are created.

Proper fix should block all distributed pre-processing due to alter table command.

**Comment by onderkalaci** [comment]
> However, this does not fix attach table since foreign key validation occurs before the table shards are created.

Could you please share steps to repro this as well?

**Comment by metdos** [comment]
> Now we have an interim fix however this is not the final fix.

@mtuncer let's continue to work on this after merging the interim fix.

**Comment by metdos** [comment]
PostgreSQL 11.2 is released, I think we should be able to have a permanent fix now.

https://www.postgresql.org/docs/11/release-11-2.html

**Comment by pykello** [comment]
> > However, this does not fix attach table since foreign key validation occurs before the table shards are created.
> 
> Could you please share steps to repro this as well?

The following use-case fails in the current master. I'm still trying to understand it.

```sql
DROP TABLE IF EXISTS partitioning_test, partitioning_test_2009, partitioning_test_2012, partitioning_test_reference CASCADE;


CREATE TABLE partitioning_test(id int, time date) PARTITION BY RANGE (time);
CREATE TABLE partitioning_test_2009 (LIKE partitioning_test);
CREATE TABLE partitioning_test_reference(id int PRIMARY KEY, subid int);

SELECT create_distributed_table('partitioning_test_2009', 'id'),
       create_distributed_table('partitioning_test', 'id'),
       create_reference_table('partitioning_test_reference');

ALTER TABLE partitioning_test ADD CONSTRAINT partitioning_reference_fkey FOREIGN KEY (id) REFERENCES partitioning_test_reference(id) ON DELETE CASCADE;

-- First ATTACH succeeds. If we DETACH, and then re-ATTACH, it fails with a deadlock.
ALTER TABLE partitioning_test ATTACH PARTITION partitioning_test_2009 FOR VALUES FROM ('2009-01-01') TO ('2010-01-01');
ALTER TABLE partitioning_test DETACH PARTITION partitioning_test_2009 ;
ALTER TABLE partitioning_test ATTACH PARTITION partitioning_test_2009 FOR VALUES FROM ('2009-01-01') TO ('2010-01-01');

ERROR:  canceling the transaction since it was involved in a distributed deadlock
```

**Comment by pykello** [comment]
@onderkalaci Can you please help me to understand why we have a deadlock at https://gist.github.com/pykello/8f562f33e25ce5c3c8b9b8b716b1d275 ? I've set `citus.distributed_deadlock_detection_factor` to -1 so I can look at the lock data.

Process 15843 is waiting for a lock that is held by process 14374. But process 14374 has all of the locks it needs, so it can continue. I am wondering why we are not committing it to unblock 15843, and instead we are reporting a distributed deadlock?

**Comment by onderkalaci** [comment]
Hey Hadi,

Let me first start with what is happening when you attach a partition to a postgres table which has a foreign key.


```SQL
DROP TABLE IF EXISTS partitioning_test, partitioning_test_2009, partitioning_test_2012, partitioning_test_reference CASCADE;

CREATE TABLE partitioning_test(id int, time date) PARTITION BY RANGE (time);
CREATE TABLE partitioning_test_2009 (LIKE partitioning_test);
CREATE TABLE partitioning_test_reference(id int PRIMARY KEY, subid int);

ALTER TABLE partitioning_test ADD CONSTRAINT partitioning_reference_fkey FOREIGN KEY (id) REFERENCES partitioning_test_reference(id) ON DELETE CASCADE;
ALTER TABLE partitioning_test_2009 ADD CONSTRAINT partitioning_reference_fkey_2009 FOREIGN KEY (id) REFERENCES partitioning_test_reference(id) ON DELETE CASCADE;

BEGIN;

	ALTER TABLE partitioning_test ATTACH PARTITION partitioning_test_2009 FOR VALUES FROM ('2009-01-01') TO ('2010-01-01');
	 SELECT relation::regclass, mode FROM pg_locks WHERE pid = pg_backend_pid() and relation = 'partitioning_test_reference'::regclass;
	          relation           |         mode          
	-----------------------------+-----------------------
	 partitioning_test_reference | AccessShareLock
	 partitioning_test_reference | RowShareLock
	 partitioning_test_reference | ShareLock
	 partitioning_test_reference | ShareRowExclusiveLock
	 partitioning_test_reference | AccessExclusiveLock
	(5 rows)

```

We see that the session acquires `AccessExclusiveLock` on the table `partitioning_test_reference`. When you do this on a distributed table, concurrent sessions initiated by `ALTER ATBLE .. ATTACH PARTITION` would acquire `AccessExclusiveLock` on the reference table on the same worker node. Thus, none of them would be able to continue and commit, forming a self-deadlock.

To handle these cases, we use sequential execution mode where all commands to a worker node is sent sequentially from a single connection. An example PR that implements sequential mode for DDLS is [here](https://github.com/citusdata/citus/pull/2189).

We've implemented/expanded sequential mode mostly for foreign keys to reference tables. Because, for several operations, parallel connections would deadlock due to reference tables. A [presentation](https://docs.google.com/presentation/d/1_CRzxKEOQriTRbbuqtnvj_nPVqfKioHz6lNHJugpq8U/edit#slide=id.p) from a team meeting which explains some more details on the topic.


We've already switch to [sequential mode](https://github.com/citusdata/citus/blob/913cac23916cb3c3f9caecc552f643897dfbb867/src/backend/distributed/commands/table.c#L1201-L1209) if DETACH partition is executed on table which has a foriegn key to reference tables.


I think we've forget to implemented it for ATTACH partition because it was not possible to attach partition with a foreign key to reference tables before pg 11.


So, the diff is probably as simple as below:

```
diff --git a/src/backend/distributed/commands/table.c b/src/backend/distributed/commands/table.c
index a687931a..9d74bd30 100644
--- a/src/backend/distributed/commands/table.c
+++ b/src/backend/distributed/commands/table.c
@@ -1198,7 +1198,7 @@ SetupExecutionModeForAlterTable(Oid relationId, AlterTableCmd *command)
                        }
                }
        }
-       else if (alterTableType == AT_DetachPartition)
+       else if (alterTableType == AT_DetachPartition || alterTableType == AT_AttachPartition)
```

**Comment by pykello** [comment]
Just for the reference, what I was missing was that because of the 2-phase commit, we cannot commit any transactions before all transactions have been prepared. So if we have two processes belonging to the same distributed transaction, and one of them is has all the locks it needs, and the 2nd process needs one of the locks the 1st process is holding, first one can not be committed because not all of the processes has been prepared, and we get a distributed deadlock.

Also I'm closing this as it is fixed by #2629.
